iam-login-service/src/main/java/it/infn/mw/iam/config/security/MitreSecurityConfig.java

```diff
@@ -180,12 +180,15 @@ public static class IntrospectEndpointAuthorizationConfig extends WebSecurityCon
     @Autowired
     @Qualifier("clientUserDetailsService")
     private UserDetailsService userDetailsService;
+    
+    @Autowired
+    private PasswordEncoder passwordEncoder;
 
     @Override
     protected void configure(final AuthenticationManagerBuilder auth) throws Exception {
 
       auth.userDetailsService(userDetailsService)
-        .passwordEncoder(NoOpPasswordEncoder.getInstance());
+        .passwordEncoder(passwordEncoder);
```

Comment:
Com fa a funzionare questa cosa? `PasswordEncoder` è un'interfaccia, non un'implementazione.
In più `passwordEncoder` è una variable non inzializzata.
Cosa fa `@Autowired`?

Reply:
Qui serve questa modifica perchè l'introspection endpoint richiede autenticazione, quindi bisogna ricalcolare l'hash del client secret.
`@Autowired` serve per iniettare automaticamente le dipendenze

Reply:
Mi fido sull'autowired, ma `private PasswordEncoder passwordEncoder;` è corretto? A meno che non quell'autowired non traformi `passwordEncoder` in un'istanza di bcrypt (ovvero `new BCryptPasswordEncoder()`) non vedo come possa funzionare.

Non è che forse dovrebbe essere

```java
@Autowired
private BCryptPasswordEncoder passwordEncoder = new BCryptPasswordEncoder();
```
?

E poiché abbiamo l'utility

```java
public static final BCryptPasswordEncoder bcrypt() {
   final short DEFAULT_ROUND = 12;    // server final?
   return new BCryptPasswordEncoder(DEFAULT_ROUND);
 }
```
utilizzerei sempre e solo questa, rinominata opportunamente.


Per concludere, possibile che diventi

```java
@Autowired
private BCryptPasswordEncoder passwordEncoder = bcrypt();
```
?